test: fix various tests to not delete resources before log gathering is performed

ianvernon opened the pull request:
A variety of tests delete resources that are created within the test (the `It` itself). This means that when the test fails, the resource is already deleted before any logs / associated metadata can be gathered about that resource. This PR attempts to refactor said tests to create said resources in `BeforeEach` and delete them in `AfterEach`.

Signed-off by: Ian Vernon <ian@cilium.io>

Relates-to: #3821

Commit:
test/k8sT: move cleanupNetworkPolicy to AfterEach within test

This ensures that resources are not deleted until after logs / debugging information is gathered
in the case that the test fails.

Signed-off by: Ian Vernon <ian@cilium.io>

Commit:
test/k8sT: wrap policy across namespaces test in Context

This will allow for creation / deletion of resources for this test to be put in `BeforeEach`
and `AfterEach` within said `Context` instead of within the `It` itself.

Signed-off by: Ian Vernon <ian@cilium.io>

Commit:
test/k8sT: move creation and deletion of resources

Move these to `BeforeEach` and `AfterEach` accordingly so logs are more
informative if test fails.

Signed-off by: Ian Vernon <ian@cilium.io>

Commit:
test/k8sT: wrap Checks Service test in `Context`

This will allow for putting creation and removal of resources within `BeforeEach`
and `AfterEach` for this test alone instead of within the `It`.

Signed-off by: Ian Vernon <ian@cilium.io>

Commit:
test/k8sT: move creation of resources outside of `It`

* Rename "Checks Service" to "Checks service on same node".
* Move creation / deletion of resources for "Checks service on same node" to
`BeforeEach` and `AfterEach`.

* Do some misc. cleanup in other tests by removing redundant type declarations,
moving all `var` declarations under `var(...)`.

Signed-off by: Ian Vernon <ian@cilium.io>

Commit:
test/k8sT: move cross-node service test within `Context`

Signed-off by: Ian Vernon <ian@cilium.io>

Commit:
test/k8sT: move creation of resources

Move creation of resources to `BeforeEach` and `AfterEach`.

Signed-off by: Ian Vernon <ian@cilium.io>

Commit:
test/k8sT: move NodePort test to within across nodes `Context`

This test uses the same resources created within the `Context`, so move it
within that `Context` to avoid useless creation / deletion of the same resource.

Signed-off by: Ian Vernon <ian@cilium.io>

eloycoto commented on test/k8sT/Policies.go:
```diff
@@ -304,68 +304,79 @@ var _ = Describe("K8sValidatedPolicyTest", func() {
-			Expect(err).Should(BeNil(), "Endpoints timeout on namespaces %q", helpers.DefaultNamespace)
+			AfterEach(func() {
+				// Explicitly do not check results to avoid incomplete teardown of test.
+				_ = kubectl.Delete(demoManifest).ExpectSuccess()
```
I think that in the AfterEach we should not use ExpectSucess. It should continue always.

ianvernon replied:
Sorry, I missed this `ExpectSuccess`. The comment above I wrote even indicates that we don't want to check the result. Will fix this in next upload.

eloycoto reviewed (commented):
minor nitpick

Commit:
test/k8sT: fix deletions in AfterEach to not have assertions

This is done to avoid incomplete teardown of the tests.

Signed-off by: Ian Vernon <ian@cilium.io>

Commit:
test/k8sT: fix instantiation of variables

Explicitly call `ManifestGet` to set variable values within `Context`.

Signed-off by: Ian Vernon <ian@cilium.io>

eloycoto commented on test/k8sT/Services.go:
```diff
@@ -112,83 +112,111 @@ var _ = Describe("K8sValidatedServicesTest", func() {
+		)
 
-		waitPodsDs()
+		BeforeEach(func() {
```
If you add in this context `BeforeAll` and `AfterAll` it'll be great

ianvernon replied:
Done, will merge once I get a green build!

eloycoto reviewed (commented):
Super minor thing, can be merged if you want ;-)

Commit:
test/k8sT: change "Checks service across nodes" to use "BeforeAll" and "AfterAll"

This will remove needless deletion and then immediate recreation of resources used
during this "Context".

Signed-off by: Ian Vernon <ian@cilium.io>
